src/json.c

```diff
@@ -0,0 +1,147 @@
+  int rv = REDISMODULE_ERR;
+  RSLanguage lang = rule->lang_default;
+  if (!rule->lang_field) {
+    goto done;
+  }
+
+  const char *langStr;
+  if (RedisJSON_GetString(jsonKey, rule->lang_field, &langStr, NULL) != REDISMODULE_OK) {
+    goto done;
+  }
+  
+  lang = RSLanguage_Find(langStr);
+  if (lang == RS_LANG_UNSUPPORTED) {
+    lang = rule->lang_default;
+    goto done;
+  }
+
+done:
+  return lang;
+}
+
+static RSLanguage SchemaRule_JsonScore(RedisModuleCtx *ctx, const SchemaRule *rule,
+                                       RedisJSONKey jsonKey, const char *keyName) {
+  double score = rule->score_default;
+  if (!rule->score_field) {
+    goto done;
+  }
+
+  if(japi && japi->getDoubleFromKey(jsonKey, rule->score_field, &score) != REDISMODULE_OK) {
+    RedisModule_Log(NULL, "warning", "invalid field %s for key %s", rule->score_field, keyName);
+  }
+
+done:
+  return score;
+}
+
+int Document_LoadSchemaFieldJson(Document *doc, RedisSearchCtx *sctx) {
+  int rv = REDISMODULE_ERR;
+  if (!japi) {
+    RedisModule_Log(sctx->redisCtx, "warning", "cannot operate on a JSON index as RedisJSON is not loaded");
+    return REDISMODULE_ERR;
+  }
+  IndexSpec *spec = sctx->spec;
+  SchemaRule *rule = spec->rule;
+  RedisModuleCtx *ctx = sctx->redisCtx;
+  size_t nitems = sctx->spec->numFields;
+  RedisJSON json = NULL;
+
+  RedisJSONKey jsonKey = japi->openKey(ctx, doc->docKey);
+  if (!jsonKey) {
+    goto done;
+  }
+  Document_MakeStringsOwner(doc); // TODO: necessary??
+
+  const char *keyName = RedisModule_StringPtrLen(doc->docKey, NULL);
+  doc->language = SchemaRule_JsonLanguage(sctx->redisCtx, rule, jsonKey, keyName);
+  doc->score = SchemaRule_JsonScore(sctx->redisCtx, rule, jsonKey, keyName);
+  // No payload on JSON as RedisJSON does not support binary fields
+
+  const char *jsonVal; //remove
+
+  size_t count;
+  JSONType type;
+  doc->fields = rm_calloc(nitems, sizeof(*doc->fields));
+  size_t ii = 0;
+  for (; ii < spec->numFields; ++ii) {
+    FieldSpec *field = &spec->fields[ii];
+
+    // retrieve json pointer
+    // TODO: check option to move to getStringFromKey
+    json = japi->get(jsonKey, field->path, &type, &count);
+    if (!json) {
+        continue;
+    }
+    if (type == JSONType_Array || type == JSONType_Object) {
+        japi->close(json);
+        json = NULL;
+      continue;
+    }
+
+    size_t oix = doc->numFields++;
+    doc->fields[oix].path = rm_strdup(field->path);
+    doc->fields[oix].name = (field->name == field->path) ? doc->fields[oix].path
+                                                         : rm_strdup(field->name);
+
+    // on crdt the return value might be the underline value, we must copy it!!!
+    // TODO: change `fs->text` to support hash or json not RedisModuleString
+    if (japi->getRedisModuleString(json, &doc->fields[oix].text) != REDISMODULE_OK) {
+      RedisModule_Log(ctx, "verbose", "Failed to load value from field %s", field->path);
+      goto done;
```

Comment:
Why would it failed here?

Reply:
Another thing we should notice is that we always get the data as a string, if the field is numeric it's not worth turn it to a string just to parse it to double again ... this optimization can be applied later...

Reply:
I do not like those silence errors, we need to open an issue to fix them if things are not supposed to happen let's assert them ... (we can do it in another PR later since we already have it on hashes anyway ...)

Reply:
Agree, but as part of this review, I would rather follow the current methodology than make changes that are unrelated to the PR.

Reply:
So let's just open an issue so it will not be forgotten...